### PR TITLE
[1LP][RFR] Added two parametrized advanced_search tests

### PR DESCRIPTION
**Pull request by mmojzis**
Added two global parametrized advanced_search tests for checking whether advanced search is possible.
Also had to add some navmazing code for tested pages.
More tests for advanced search/filters are on the way.

### CODE REVIEW

**Comment by dajoRH** [comment]
Lint report for commit ce67e45dfab0cc0704fb845bda1c57b4b55245f4:

`cfme/cloud/instance/__init__.py`:
- :bangbang: Line 21:101: **E501** *line too long (106 > 100 characters)*

`cfme/networks/views.py`:
- :bangbang: Line 8:77: **E231** *missing whitespace after ','*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 0c4fed4a7c58162ff6eecc02124a68a1a4f841df:

`cfme/tests/webui/test_advanced_search.py`:
- :bangbang: Line 95:101: **E501** *line too long (186 > 100 characters)*
- :bangbang: Line 95:138: **E231** *missing whitespace after ','*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit 92a873712e199177122cd199278a355fe7bf01f6:

`cfme/tests/webui/test_advanced_search.py`:
- :heavy_exclamation_mark: Line 21:1: **W293** *blank line contains whitespace*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

**Comment by mmojzis** [comment]
Tests in location "containers" are failing because is_displayed method is broken for this page. AFAIK it's going to be fixed soon.

File: cfme/ansible_tower/explorer.py [hunk 1]
```diff
@@ -0,0 +1,149 @@
...
+
+
+@navigator.register(TowerExplorerProvider, 'All')
+class TowerExplorerProviderAll(CFMENavigateStep):
+    VIEW = TowerExplorerProviderDefaultView
+    prerequisite = NavigateToAttribute('appliance.server', 'LoggedIn')
+
+    def step(self, *args, **kwargs):
+        self.prerequisite_view.navigation.select('Automation', 'Ansible Tower', 'Explorer')
+        self.view.search.clear_simple_search()
```
[the remark this answers is not in the document]
Comment (mmojzis) [reply cfme/ansible_tower/explorer.py]: You mean whole step?

[the remark this answers is not in the document]
Comment (mmojzis) [reply cfme/ansible_tower/explorer.py]: Deleted

File: cfme/ansible_tower/jobs.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
+from navmazing import NavigateToAttribute
+from widgetastic.widget import Text, View
+from widgetastic_patternfly import Dropdown
+
+from cfme.base.login import BaseLoggedInPage
+from cfme.utils.appliance import Navigatable
+from cfme.utils.appliance.implementations.ui import navigator, CFMENavigateStep
+from widgetastic_manageiq import Search, ItemsToolBarViewSelector
```
[the remark this answers is not in the document]
Comment (mmojzis) [reply cfme/ansible_tower/jobs.py]: Fixed

File: cfme/tests/webui/test_advanced_search.py [hunk 3]
```diff
@@ -0,0 +1,121 @@
...
+        view = navigate_to(getattr(appliance.collections, params[0]), params[1])
+    else:
+        view = navigate_to(params[0], params[1])
+    return view
+
+
+def test_advanced_search_button_displayed(params, appliance):
+    view = _navigation(params, appliance)
+    if not view.search.is_advanced_search_possible:
+        raise pytest.fail(
```
[the remark this answers is not in the document]
Comment (mmojzis) [reply cfme/tests/webui/test_advanced_search.py]: Fixed.

File: cfme/services/workloads.py [hunk 4]
```diff
@@ -108,9 +108,9 @@ def is_displayed(self):
                 "All Templates & Images"])
 
 
-class BaseWorkloads(NavigatableMixin):
-    def __init__(self, appliance):
-        self.appliance = appliance
+class BaseWorkloads(Navigatable):
```
Comment (quarckster) [inline cfme/services/workloads.py]: **QUESTION**
Why did you change that?

File: cfme/ansible_tower/explorer.py [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+class TowerExplorerProvider(Navigatable):
+    pass
+
+
+class TowerExplorerSystem(Navigatable):
+    pass
+
+
+class TowerExplorerJobTemplates(Navigatable):
+    pass
```
Comment (quarckster) [inline cfme/ansible_tower/explorer.py]: **REQUIRED**
All these objects should have a respectful collection.

File: cfme/ansible_tower/explorer.py [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+
+
+@navigator.register(TowerExplorerProvider, 'All')
+class TowerExplorerProviderAll(CFMENavigateStep):
+    VIEW = TowerExplorerProviderDefaultView
+    prerequisite = NavigateToAttribute('appliance.server', 'LoggedIn')
+
+    def step(self, *args, **kwargs):
+        self.prerequisite_view.navigation.select('Automation', 'Ansible Tower', 'Explorer')
```
Comment (quarckster) [inline cfme/ansible_tower/explorer.py]: **REQUIRED**
`self.prerequisite_view.navigation.select('Automation', 'Ansible Tower', 'Explorer')` should be a separate destination, let's say `TowerExplorer`:
```python
@navigator.register(Server)
class TowerExplorer(CFMENavigateStep):
    VIEW = TowerExplorerView
    prerequisite = NavigateToSibling('LoggedIn')

    def step(self):
        self.view.navigation.select('Automation', 'Ansible Tower', 'Explorer')
```
Other destinations should use it as a prerequisite.

File: cfme/cloud/host_aggregates.py [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    @property
+    def is_displayed(self):
+        return (
+            self.in_host_aggregates and
+            self.title.text == 'Host Aggregates'
+        )
+
+
+class HostAggregates(Navigatable):
```
Comment (quarckster) [inline cfme/cloud/host_aggregates.py]: **REQUIRED**
Please make a collection here.

File: cfme/containers/build.py [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+    @property
+    def is_displayed(self):
+        return (
+            self.in_build and
+            self.title.text == 'Container Builds'
+        )
+
+
+class Build(Navigatable):
```
Comment (quarckster) [inline cfme/containers/build.py]: **REQUIRED**
^

File: cfme/services/workloads.py (hunk 4 above)
Comment (mmojzis) [reply cfme/services/workloads.py]: Because I was told that NavigatableMixin is deprecated and Navigatable should be used instead.

File: cfme/ansible_tower/explorer.py (hunk 5 above)
Comment (mmojzis) [reply cfme/ansible_tower/explorer.py]: I don't know why. I don't need collection for my tests and I would say same of others as nobody created collection for that until now. Also I have never worked with collections and I don't have time now to learn that...I just need to have these tests to have whole FA Automated.

File: cfme/ansible_tower/explorer.py [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+        tree = ManageIQTree()
+
+    @View.nested
+    class tower_explorer_job_templates(Accordion):  # noqa
+        ACCORDION_NAME = 'Job Templates'
+        tree = ManageIQTree()
+
+
+class TowerExplorerProviderToolbar(View):
+    reload = Button(title=VersionPick({Version.lowest(): 'Reload current display',
```
Comment (digitronik) [inline cfme/ansible_tower/explorer.py]: Please, do not use `VersionPick`. we already removed it from repo.

File: cfme/ansible_tower/explorer.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+                                       '5.9': 'Refresh this page'}))
+    configuration = Dropdown('Configuration')
+    policy = Dropdown('Policy')
+    download = Dropdown('Download')
+    view_selector = View.nested(ItemsToolBarViewSelector)
+
+
+class TowerExplorerSystemJobTemplatesToolbar(View):
+    reload = Button(title=VersionPick({Version.lowest(): 'Reload current display',
```
Comment (digitronik) [inline cfme/ansible_tower/explorer.py]: same `VersionPick`

File: cfme/ansible_tower/explorer.py (hunk 9 above)
Comment (mmojzis) [reply cfme/ansible_tower/explorer.py]: Okay, what should I use instead?

Comment (digitronik) [reply cfme/ansible_tower/explorer.py]: my mistook, ignore it. I forgot we are playing with `View`. WT `View` classes not supporting `contextual object` access. Yes, we can use `browser.appliance.version` mainly pointing to `appliance.version` in `methods` of `View` class but for `widget` initialization not possible with it. Same thing applicable to `ParametrizeView` :D

File: cfme/services/workloads.py (hunk 4 above)
Comment (mmojzis) [reply cfme/services/workloads.py]: @quarckster Should I change this back?

**Comment by dajoRH** [comment]
Would you mind rebasing this Pull Request against latest master, please? :trollface:
*CFME QE Bot*

**Comment by dajoRH** [comment]
Lint report for commit daeebdc0bbe1171d4764a52c98a25afa1d7dca77:

`cfme/storage/volume.py`:
- :red_circle: Line 8:1: **F811** *redefinition of unused 'Button' from line 6*
- :red_circle: Line 8:1: **F811** *redefinition of unused 'Dropdown' from line 6*

Please, rectify these issues :smirk: .


**No commit flaws detected**.

*CFME QE Bot*

File: cfme/services/workloads.py (hunk 4 above)
Comment (quarckster) [reply cfme/services/workloads.py]: You can keep it.

File: cfme/storage/volume.py [hunk 11]
```diff
@@ -4,6 +4,9 @@
 from navmazing import NavigateToSibling, NavigateToAttribute
 from widgetastic.widget import TextInput
 from widgetastic_patternfly import BreadCrumb, Button, Dropdown
+from widgetastic_manageiq import Search
+from widgetastic_patternfly import Button, Dropdown
```
Comment (quarckster) [inline cfme/storage/volume.py]: **REQUIRED**
`from widgetastic_patternfly import Button, Dropdown` is already imported.

File: setup.py [hunk 12]
```diff
@@ -88,6 +88,11 @@
             'ansible_credentials = cfme.ansible.credentials:CredentialsCollection',
             'ansible_playbooks = cfme.ansible.playbooks:PlaybooksCollection',
             'ansible_repositories = cfme.ansible.repositories:RepositoryCollection',
+            ('ansible_tower_job_templates = '
+                'cfme.ansible_tower.explorer:AnsibleTowerJobTemplatesCollection'),
+            'ansible_tower_providers = cfme.ansible_tower.explorer:AnsibleTowerProvidersCollection',
+            'ansible_tower_systems = cfme.ansible_tower.explorer:AnsibleTowerSystemsCollection',
+            'ansible_tower_jobs = cfme.ansible_tower.jobs:TowerJobsCollection',
```
Comment (quarckster) [inline setup.py]: **REQUIRED**
Please keep alphabetical order.

File: cfme/storage/volume.py (hunk 11 above)
Comment (mmojzis) [reply cfme/storage/volume.py]: Oops, I see...Rebasing 20 files easily creates a mistake.
Fixed.

File: setup.py (hunk 12 above)
Comment (mmojzis) [reply setup.py]: Fixed.

File: cfme/tests/webui/test_advanced_search.py [hunk 13]
```diff
@@ -0,0 +1,113 @@
...
+     'Configured System (Ansible Tower) : Hostname'),
+    ('ansible_tower_job_templates', 'All', 'ansible_tower_explorer_job_templates',
+     'Job Template (Ansible Tower) : Name'),
+    ('ansible_tower_jobs', 'All', 'ansible_tower_jobs', 'Ansible Tower Job : Name'),
+]
+
+pytestmark = [
+    pytest.mark.parametrize(
+        'params', params_values,
+        ids=['{}_{}'.format(param[2], param[1].lower()) for param in params_values]
```
Comment (psav) [inline cfme/tests/webui/test_advanced_search.py]: **REQUEST:** Can we please have ```-``` between the ids, it is the default for separating parameters.

Comment (mmojzis) [reply cfme/tests/webui/test_advanced_search.py]: Done
